### PR TITLE
feat(useWebNotification): new function

### PR DESCRIPTION
**Pull request by michael-observerly**

### CODE REVIEW

**Comment by michael-observerly** [comment]
Linked Issue: #975

**Comment by michael-observerly** [comment]
@antfu Would you like an accompanying test to go with this new feature? (Hope that is not a stupid question? :D)

File: packages/core/useWebNotification/index.ts [hunk 1]
```diff
@@ -0,0 +1,120 @@
+import { onMounted, Ref, ref, onUnmounted } from 'vue-demi'
+
+interface WebNotificationOptions {
+  /*
+   *
+   * The body string of the notification as specified in the constructor's options parameter.
```
Comment (antfu) [inline packages/core/useWebNotification/index.ts:6]: Can you follow the TSDocs format from other composables and remove the extract spaces?

Comment (michael-observerly) [reply packages/core/useWebNotification/index.ts:6]: Will do. Will push shortly.

**Comment by michael-observerly** [comment]
@antfu Amends pushed. Thanks for the super quick review.

**Comment by michael-observerly** [comment]
@antfu I've spotted a few typos, I'll fix these shortly.

**Comment by michael-observerly** [comment]
@antfu Some major and minor changes have been pushed, will require further review. Thanks!

**Comment by antfu** [comment]
I will see it, there is no need to ping me :)

File: packages/core/useWebNotification/index.ts [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+  // Is the web notifications API supported?:
+  const isSupported: boolean = 'Notification' in window
+
+  // Request permission to use web notifications:
+  const requestPermission = async() => {
+    if ('permission' in Notification && Notification.permission !== 'denied') await Notification.requestPermission()
+  }
+
+  // Show notification method:
+  const show = (): void => {
```
Comment (antfu) [inline packages/core/useWebNotification/index.ts:125]:
```suggestion
  const show = (options?: WebNotificationOptions): void => {
```

I think it's better to allow passing an option. The top-level option could be the default for merging.

File: packages/core/useWebNotification/index.ts [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+ * Reactive useWebNotification
+ *
+ * @see https://vueuse.org/useWebNotification
+ * @param title
+ * @param options of type WebNotificationOptions
+ * @param methods of type WebNotificationMethods
+ */
+export const useWebNotification = (
+  title: string,
+  options: WebNotificationOptions = {},
```
Comment (antfu) [inline packages/core/useWebNotification/index.ts:111]: I feel we could merge the title and options into one argument.

File: packages/core/useWebNotification/index.ts [hunk 4]
```diff
@@ -0,0 +1,172 @@
+import { onMounted, ref, onUnmounted } from 'vue-demi'
+import type { Ref } from 'vue-demi'
+
+import { useEventListener } from '../useEventListener'
+
+interface WebNotificationOptions {
```
Comment (antfu) [inline packages/core/useWebNotification/index.ts:6]:
```suggestion
export interface UseWebNotificationOptions {
```

File: packages/core/useWebNotification/index.ts [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+  /**
+   * A handler for the show event. It is triggered when the notification is displayed.
+   *
+   * @default null
+   */
+  onShow: ((e: Event) => void) | null
+}
+
+// Default web notification methods:
+const defaultWebNotificationMethods = {
```
Comment (antfu) [inline packages/core/useWebNotification/index.ts:94]: For hooks, we could use https://vueuse.org/shared/createeventhook/

**Review by antfu — CHANGES_REQUESTED** [review]

File: packages/core/useWebNotification/index.ts (hunk 3 above)
Comment (michael-observerly) [reply packages/core/useWebNotification/index.ts]: Added amends to this PR.

File: packages/core/useWebNotification/index.ts (hunk 2 above)
Comment (michael-observerly) [reply packages/core/useWebNotification/index.ts]: Added amends to this PR.

File: packages/core/useWebNotification/index.ts (hunk 5 above)
Comment (michael-observerly) [reply packages/core/useWebNotification/index.ts:110]: @antfu Could you give an example of how you would like this implemented?

Comment (michael-observerly) [reply packages/core/useWebNotification/index.ts]: @antfu I've added my interpretation into commit 1f01933. Thank you for the suggestions. :)

Comment (michael-observerly) [reply packages/core/useWebNotification/index.ts]: * Includes usage documentation changes as well.

**Comment by michael-observerly** [comment]
@antfu This one seems to have not come through onto vueuse.org documentation...

**Pull request merged**